{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8367d500",
   "metadata": {},
   "source": [
    "## Sequence Features\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b339c53c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "# Function to remove letter 'O' from a sequence\n",
    "def remove_letter_O(sequence):\n",
    "    return sequence.replace('O', '')\n",
    "\n",
    "# Function to extract k-mer frequencies\n",
    "def kmer_frequencies(seq, k):\n",
    "    kmers = {}\n",
    "    for i in range(len(seq) - k + 1):\n",
    "        kmer = seq[i:i+k]\n",
    "        kmers[kmer] = kmers.get(kmer, 0) + 1\n",
    "    total_kmers = sum(kmers.values())\n",
    "    for kmer, count in kmers.items():\n",
    "        kmers[kmer] = count / total_kmers\n",
    "    return kmers\n",
    "\n",
    "# Input FASTA file\n",
    "fasta_file = \"all_lncrna_location_1057.fasta\"\n",
    "\n",
    "# Parameter for k-mer length\n",
    "#k = 2  # Change this value to the desired k-mer length\n",
    "\n",
    "# Initialize a list to store feature data\n",
    "feature_data = []\n",
    "\n",
    "\n",
    "# Loop through each sequence in the FASTA file\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    sequence_id = record.id\n",
    "    sequence = str(record.seq)\n",
    "    \n",
    "    # Remove 'O' from the sequence\n",
    "    sequence = remove_letter_O(sequence)\n",
    "    \n",
    "    # Initialize a dictionary to store features for this sequence\n",
    "    features = {}\n",
    "    \n",
    "    # Composition\n",
    "    for base in ['A', 'T', 'C', 'G']:\n",
    "        features[f\"{base}_freq\"] = sequence.count(base) / len(sequence)\n",
    "    \n",
    "    # GC content\n",
    "    features[\"GC_content\"] = (sequence.count('G') + sequence.count('C')) / len(sequence)\n",
    "    \n",
    "    # k-mer frequencies, dimer\n",
    "    kmers = kmer_frequencies(sequence, 2)\n",
    "    for kmer, freq in kmers.items():\n",
    "        features[f\"{kmer}_freq\"] = freq\n",
    "        \n",
    "    # k-mer frequencies, trimer\n",
    "    kmers = kmer_frequencies(sequence, 3)\n",
    "    for kmer, freq in kmers.items():\n",
    "        features[f\"{kmer}_freq\"] = freq\n",
    "        \n",
    "    # k-mer frequencies, tetramer\n",
    "    kmers = kmer_frequencies(sequence, 4)\n",
    "    for kmer, freq in kmers.items():\n",
    "        features[f\"{kmer}_freq\"] = freq\n",
    "        \n",
    "    # k-mer frequencies,pentamer\n",
    "    kmers = kmer_frequencies(sequence, 5)\n",
    "    for kmer, freq in kmers.items():\n",
    "        features[f\"{kmer}_freq\"] = freq\n",
    "        \n",
    "    # k-mer frequencies,pentamer\n",
    "    kmers = kmer_frequencies(sequence, 6)\n",
    "    for kmer, freq in kmers.items():\n",
    "        features[f\"{kmer}_freq\"] = freq\n",
    "        \n",
    "    # Add sequence ID and class (you may need to modify this part)\n",
    "    # For example, if your FASTA header contains class information:\n",
    "    # class_info = sequence_id.split(\"|\")[1]\n",
    "    # features[\"Class\"] = class_info\n",
    "    \n",
    "    feature_data.append(features)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d31ff05",
   "metadata": {},
   "source": [
    "## Structure features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7a08081b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "def parse_rnafold_output_enhanced(filename):\n",
    "    with open(filename, 'r') as f:\n",
    "        lines = f.readlines()\n",
    "    features_list = []\n",
    "    for i in range(0, len(lines), 1):\n",
    "        # Ensure we have at least two lines to process\n",
    "        if i + 1 >= len(lines):\n",
    "            break\n",
    "        # Get the sequence and structure info\n",
    "        seq = lines[i].strip()\n",
    "        # Check if the next line contains the expected structure format\n",
    "        if '(' in lines[i+1] and ')' in lines[i+1]:\n",
    "            struct_info = lines[i+1].strip().split()\n",
    "            if len(struct_info) < 2:\n",
    "                continue\n",
    "            struct = struct_info[0]\n",
    "            mfe = float(struct_info[1].replace('(', '').replace(')', ''))\n",
    "            # Compute features\n",
    "            total_bases = len(struct)\n",
    "            loop_percentage = struct.count('.') / total_bases\n",
    "            paired_percentage = (struct.count('(') + struct.count(')')) / total_bases\n",
    "\n",
    "            # Count occurrences of hairpins, bulges, and internal loops\n",
    "            hairpin_loops = len(re.findall('\\(\\.+?\\)', struct)) \n",
    "            hairpin_loops_percentage = len(re.findall('\\(\\.+?\\)', struct)) / total_bases\n",
    "            bulges = len(re.findall('\\(\\.+\\)|\\)\\.+?\\(', struct))\n",
    "            bulges_percentage = len(re.findall('\\(\\.+\\)|\\)\\.+?\\(', struct)) / total_bases\n",
    "            internal_loops = len(re.findall('\\(\\..+?\\.\\)', struct))\n",
    "            internal_loops_percentage = len(re.findall('\\(\\..+?\\.\\)', struct)) / total_bases\n",
    "\n",
    "            features = {\n",
    "                'Total_length': total_bases,\n",
    "                'MFE': mfe,\n",
    "                'Loop_percentage': loop_percentage,\n",
    "                'Stem_percentage': paired_percentage,\n",
    "                'Hairpin_loops': hairpin_loops,\n",
    "                'Hairpin_loops_percentage': hairpin_loops_percentage,\n",
    "                'Bulges': bulges,\n",
    "                'Bulges_percentage': bulges_percentage,\n",
    "                'Internal_loops': internal_loops,\n",
    "                'Internal_loops_percentage': internal_loops_percentage\n",
    "                \n",
    "            }\n",
    "\n",
    "            features_list.append(features)\n",
    "\n",
    "    return features_list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "436c19c2",
   "metadata": {},
   "source": [
    "## Triple helix motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dfb5afd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "def has_triple_helix_motif(sequence):\n",
    "    # Search for a run of purines at the 3' end\n",
    "    purine_run = re.search('[AG]{5,}$', sequence)\n",
    "    return 1 if purine_run else 0\n",
    "\n",
    "def extract_triple_helix_feature_from_fasta(input_file):\n",
    "    features = {}\n",
    "    with open(input_file, 'r') as f:\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "            sequence = str(record.seq)\n",
    "            features[record.id] = has_triple_helix_motif(sequence)\n",
    "    return features\n",
    "\n",
    "# Path to the provided multi-FASTA file\n",
    "input_file_path = \"all_lncrna_location_1057.fasta\"\n",
    "\n",
    "# Extract features from the provided file\n",
    "triple_helix_features = extract_triple_helix_feature_from_fasta(input_file_path)\n",
    "\n",
    "# Convert the features dictionary to a DataFrame\n",
    "df_triple_helix = pd.DataFrame(list(triple_helix_features.items()), columns=[\"Sequence_ID\", \"Triple_Helix_Motif\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "391a307e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "def count_palindromes_by_length(sequence, min_length=12, max_length=20):\n",
    "   \n",
    "    sequence = str(sequence).upper()\n",
    "    palindrome_counts = {i: 0 for i in range(min_length, max_length+1)}\n",
    "    \n",
    "    for length in range(min_length, max_length + 1):\n",
    "        for start in range(0, len(sequence) - length + 1):\n",
    "            segment = sequence[start:start+length]\n",
    "            if segment == segment[::-1]:\n",
    "                palindrome_counts[length] += 1\n",
    "                \n",
    "    return palindrome_counts\n",
    "\n",
    "def extract_palindrome_counts_from_fasta(input_file, min_length=12, max_length=20):\n",
    "    data = []\n",
    "    for record in SeqIO.parse(input_file, \"fasta\"):\n",
    "        sequence = record.seq\n",
    "        palindrome_counts = count_palindromes_by_length(sequence, min_length, max_length)\n",
    "        data.append({\n",
    "            \"Sequence_ID\": record.id,\n",
    "            **palindrome_counts\n",
    "        })\n",
    "    return pd.DataFrame(data)\n",
    "\n",
    "input_file_path = \"all_lncrna_location_1057.fasta\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018dcd08",
   "metadata": {},
   "source": [
    "### ORF Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c90bd09b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "\n",
    "# Define the find_longest_orf function again\n",
    "def find_longest_orf(sequence):\n",
    "    \"\"\"Find the longest ORF in a given sequence.\"\"\"\n",
    "    start_codon = 'ATG'\n",
    "    stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "    longest_orf = None\n",
    "    orf_count = 0\n",
    "\n",
    "    for frame in range(3):\n",
    "        for i in range(frame, len(sequence), 3):\n",
    "            if sequence[i:i+3] == start_codon:\n",
    "                for j in range(i+3, len(sequence), 3):\n",
    "                    if sequence[j:j+3] in stop_codons:\n",
    "                        orf_count += 1\n",
    "                        if longest_orf is None or (j + 3 - i) > len(longest_orf[0]):\n",
    "                            longest_orf = (sequence[i:j+3], i, j+3, frame)\n",
    "                        break\n",
    "    return longest_orf, orf_count\n",
    "\n",
    "# Modify the code to add a default set of values for sequences without ORFs\n",
    "fasta_file = \"all_lncrna_location_1057.fasta\"\n",
    "feature_data = []\n",
    "\n",
    "for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "    sequence = str(record.seq)\n",
    "    sequence = sequence.replace('O', '')\n",
    "    \n",
    "    longest_orf, orf_count = find_longest_orf(sequence)\n",
    "\n",
    "    if longest_orf is not None:\n",
    "        orf_seq, start, end, frame = longest_orf\n",
    "        orf_length = len(orf_seq)\n",
    "        gc_content = (orf_seq.count('G') + orf_seq.count('C')) / orf_length\n",
    "        amino_acid_seq = Seq(orf_seq).translate()\n",
    "\n",
    "        features = {\n",
    "            'Sequence_ID': record.id,\n",
    "            'ORF_Length': orf_length,\n",
    "            'ORF_GC_Content': gc_content,\n",
    "            'ORF_Count': orf_count,\n",
    "        }\n",
    "    else:\n",
    "        # Default values for sequences without ORFs\n",
    "        features = {\n",
    "            'Sequence_ID': record.id,\n",
    "            'ORF_Length': 0,\n",
    "            'ORF_GC_Content': 0,\n",
    "            'ORF_Count': 0,\n",
    "        }\n",
    "        \n",
    "    feature_data.append(features)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6549bc49",
   "metadata": {},
   "source": [
    "## Physico-chemical Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2c57a156",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "def calculate_properties(sequence):\n",
    "    \"\"\"Calculate molecular weight, GC content, and Tm for an RNA sequence.\"\"\"\n",
    "    # Molecular weights of individual nucleotides in g/mol\n",
    "    mw = {\n",
    "        'A': 347.2212, \n",
    "        'U': 324.1813, \n",
    "        'G': 363.2206, \n",
    "        'C': 323.1965\n",
    "    }\n",
    "    \n",
    "    # Calculate molecular weight of the sequence\n",
    "    sequence_mw = sum([mw.get(nucleotide, 0) for nucleotide in sequence]) - (len(sequence) - 1) * 79.0  # subtracting phosphate weight\n",
    "    # Calculate GC content\n",
    "    gc_content = (sequence.count('G') + sequence.count('C')) / len(sequence)\n",
    "    # Calculate melting temperature (Tm)\n",
    "    tm = 81.5 + 16.6 * 0.05 + 0.41 * gc_content * 100 - 600 / len(sequence)  # assuming 50mM Na+\n",
    "    return sequence_mw, gc_content * 100, tm\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8676833",
   "metadata": {},
   "source": [
    "## Physico_chemical Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b7567eb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main function to extract properties from the FASTA file and save to CSV\n",
    "def calculate_properties_for_fasta(input_fasta, output_csv):\n",
    "    data = []\n",
    "\n",
    "    for record in SeqIO.parse(input_fasta, \"fasta\"):\n",
    "        sequence = str(record.seq).upper()\n",
    "        mw, gc, tm = calculate_properties(sequence)\n",
    "        data.append({\n",
    "            \"Sequence_ID\": record.id,\n",
    "            \"Molecular_Weight_g/mol\": mw,\n",
    "            \"GC_Content_%\": gc,\n",
    "            \"Melting_Temperature_C\": tm\n",
    "        })\n",
    "\n",
    "    df = pd.DataFrame(data)\n",
    "    df.to_csv(output_csv, index=False)\n",
    "    \n",
    "    return df\n",
    "# Extract properties from the provided FASTA file and save to CSV\n",
    "input_fasta_path = \"all_lncrna_location_1057.fasta\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
